File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int mini = *min_element(nums.begin(),nums.end());
        if(mini>0){
            int product = 1;
            for(int i = 0 ; i < nums.size() ; i++){
                product = product * nums[i];
            }
            return product;
        }
        int product = 1;
        int maxproduct = INT_MIN;

            for(int i = 0 ; i < nums.size() ; i++){
                product = 1;
                for(int j = i ; j<nums.size();j++){
                    product *= nums[j];
                    maxproduct = max(product,maxproduct);
                }
            }

        return maxproduct;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: list[int]) -> int:
        pre = 1
        suff = 1
        ans = -2**31
        for i in range(len(nums)):
            
            if pre == 0: pre = 1
            if suff == 0: suff = 1

            pre *= nums[i]
            suff *= nums[len(nums)-(i+1)]

            ans = max(ans,suff,pre)
        
        return ans
